swap-feed/src/kucoin.rs

```diff
@@ -0,0 +1,424 @@
+        let ping_interval = Duration::from_millis(ping_interval_ms);
+        tracing::debug!(%ws_url, ?ping_interval);
+
+        let (rate_stream, _) = tokio_tungstenite::connect_async(ws_url)
+            .await
+            .context("Failed to connect to KuCoin websocket API")?;
+
+        let (to_kucoin, rate_stream) = rate_stream.split();
+        let to_kucoin = Arc::new(Mutex::new(to_kucoin));
+
+        let stream = rate_stream
+            .err_into()
+            .try_filter_map(move |msg| parse_message(msg, to_kucoin.clone(), ping_interval))
+            .boxed();
+
+        Ok(stream)
+    }
+
+    type ToKucoin = futures::stream::SplitSink<
+        tokio_tungstenite::WebSocketStream<
+            tokio_tungstenite::MaybeTlsStream<tokio::net::TcpStream>,
+        >,
+        tungstenite::Message,
+    >;
+
+    /// Parse a websocket message into a [`wire::PriceUpdate`].
+    ///
+    /// Messages which are not actually ticker updates are ignored and result in
+    /// `None` being returned. In the context of a [`TryStream`], these will
+    /// simply be filtered out.
+    async fn parse_message(
+        msg: tungstenite::Message,
+        to_kucoin: Arc<Mutex<ToKucoin>>,
+        ping_interval: Duration,
+    ) -> Result<Option<wire::PriceUpdate>, Error> {
+        let msg = match msg {
+            tungstenite::Message::Text(msg) => msg,
+            tungstenite::Message::Close(close_frame) => {
+                if let Some(tungstenite::protocol::CloseFrame { code, reason }) = close_frame {
+                    tracing::error!(
+                        "KuCoin rate stream was closed with code {} and reason: {}",
+                        code,
+                        reason
+                    );
+                } else {
+                    tracing::error!("KuCoin rate stream was closed without code and reason");
+                }
+
+                return Err(Error::ConnectionClosed);
+            }
+            msg => {
+                tracing::trace!(
+                    "KuCoin rate stream returned non text message that will be ignored: {}",
+                    msg
+                );
+
+                return Ok(None);
+            }
+        };
+
+        let update = match serde_json::from_str::<wire::Event>(&msg) {
+            Ok(wire::Event::Welcome) => {
+                tracing::debug!("Connected to KuCoin websocket API");
+
+                to_kucoin
+                    .lock()
+                    .await
+                    .send(SUBSCRIBE_XMR_BTC_TICKER_PAYLOAD.into())
+                    .await?;
+                tokio::spawn(async move {
+                    let mut ping_timer = tokio::time::interval_at(
+                        tokio::time::Instant::now() + ping_interval,
+                        ping_interval,
+                    );
+                    loop {
+                        ping_timer.tick().await;
+                        tracing::debug!("Renewing lease");
```

Comment:
same here, either add more context or a tracing macro (same goes for the log message below). I prefer the instrument macro as it's less repetetive.